src/control-plane/backend/lambda/api/service/reporting.ts

```diff
@@ -249,131 +193,315 @@ export class ReportingServ {
-          dashboardVersion: Number.parseInt(versionNumber!),
-          analysisId: query.analysisId,
-          analysisArn: newAnalysis.Arn!,
-          analysisName: query.analysisName,
-          visualIds: [visualId, tableVisualId],
-        };
-
-        logger.info('funnel chart successfully updated');
+      const sqls = [sql];
+      sqls.push(`grant select on ${query.appId}.${viewName} to ${dashboardCreateParameters.quickSight.redshiftUser}`);
+
+      const datasetPropsArray: DataSetProps[] = [];
+      datasetPropsArray.push({
+        name: '',
+        tableName: viewName,
+        columns: funnelVisualColumns,
+        importMode: 'DIRECT_QUERY',
+        customSql: `select * from ${query.appId}.${viewName}`,
+        projectedColumns: [
+          'event_date',
+          'event_name',
+          'x_id',
+        ],
+      });
+
+      let sheetId;
+      if (!query.dashboardId) {
+        sheetId = uuidv4();
+      } else {
+        if (!query.sheetId) {
+          return res.status(400).send(new ApiFail('missing required parameter sheetId'));
+        }
+        sheetId = query.sheetId;
       }
 
+      const visualId = uuidv4();
+      const visualDef = getEventLineChartVisualDef(visualId, viewName, query.timeUnit);
+      const visualRelatedParams = getVisualRelatedDefs({
+        timeScopeType: query.timeScopeType,
+        sheetId,
+        visualId,
+        viewName,
+        lastN: query.lastN,
+        timeUnit: query.timeUnit,
+        timeStart: query.timeStart,
+        timeEnd: query.timeEnd,
+      });
+
+      const visualProps = {
+        sheetId: sheetId,
+        visual: visualDef,
+        dataSetIdentifierDeclaration: [],
+        filterControl: visualRelatedParams.filterControl,
+        parameterDeclarations: visualRelatedParams.parameterDeclarations,
+        filterGroup: visualRelatedParams.filterGroup,
+        eventCount: query.eventAndConditions.length,
+      };
+
+      const tableVisualId = uuidv4();
+      const tableVisualDef = getEventPivotTableVisualDef(tableVisualId, viewName, query.groupColumn);
+
+      visualRelatedParams.filterGroup!.ScopeConfiguration!.SelectedSheets!.SheetVisualScopingConfigurations![0].VisualIds!.push(tableVisualId);
+
+      const tableVisualProps = {
+        sheetId: sheetId,
+        visual: tableVisualDef,
+        dataSetIdentifierDeclaration: [],
+      };
+
+      const result: CreateDashboardResult = await this.create(viewName, query, sqls, datasetPropsArray, [visualProps, tableVisualProps]);
+      result.visualIds.push({
+        name: 'CHART',
+        id: visualId,
+      });
+      result.visualIds.push({
+        name: 'TABLE',
+        id: tableVisualId,
+      });
+
       return res.status(201).json(new ApiSuccess(result));
     } catch (error) {
       next(error);
     }
   };
 
+  private async create(resourceName: string, query: any, sqls: string[], datasetPropsArray: DataSetProps[], visualPropsArray: VisualProps[]) {
```

Comment:
It's a very long procedure for creating or updating the dashboard, how do you handle the error if it happens in any intermittent step?

Reply:
will deal with this problem later, #149